File: MotionParallax/Frame.java
package MotionParallax;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;

import MotionParallax.Utils;
import MotionParallax.ObjDetection;
import MotionParallax.Point;

class Frame {

    protected static long _next_frame_number = 0;

    protected double camera_bearing;

    public Point camera_point;

    protected double [] detection_bearings;
    protected HashMap<Double, ObjDetection> detection_objects = new HashMap<>();

    public Frame prev;
    public long frame_number;

    public final double THRESHOLD1 = Math.PI / 8.0d;
    public final double THRESHOLD2 = Math.PI / 16.0d;
    public final int MAXLISTLENGTH = 80;

    Frame(double camera_bearing, double x, double z, double[] relative_bearings, Frame prev) {
        frame_number = _next_frame_number;
        _next_frame_number++;

        if (frame_number >= MAXLISTLENGTH) {
            Frame cur = this;
            while (cur.prev.prev != null) {
                cur = cur.prev;
            }
            cur.prev.clearChildren();
            cur.prev.camera_point = null;
            cur.prev = null;
        }

        this.camera_bearing = camera_bearing;
        this.camera_point = new Point(x, z);
        double[] absolute_bearings = new double[relative_bearings.length];
        for (int i = 0; i < absolute_bearings.length; i++) {
            double d = relative_bearings[i];
            absolute_bearings[i] = Utils.normalized(camera_bearing + d);
        }
        this.detection_bearings = absolute_bearings;
        this.prev = prev;
        genDetectionObjects();
    }

    protected void genDetectionObjects() {
        for (int i = 0; i < detection_bearings.length; i++) {
            detection_objects.put(detection_bearings[i], new ObjDetection(this, detection_bearings[i], null));
        }
    }

    public double x() {
        return camera_point.x();
    }

    public double z() {
        return camera_point.z();
    }

    public double getCameraBearing() {
        return camera_bearing;
    }

    public double[] getDetectionBearings() {
        return detection_bearings;
    }

    public HashMap<Double, ObjDetection> getDetectionObjects() {
        return detection_objects;
    }

    public double bearingToFrame(Frame other) {
        double deltaz = other.z() - z();
        double deltax = other.x() - x();
        return Math.atan2(deltaz, deltax);
    }

    public double distToFrame(Frame other) {
        double deltaz = other.z() - z();
        double deltax = other.x() - x();
        return Math.sqrt(deltaz*deltaz + deltax*deltax);
    }

    public void clearChildren() {
        for (double aBearing : getDetectionBearings()) {
            ObjDetection anObj = detection_objects.get(aBearing);
            if (anObj != null) {
                anObj.prev = null;
                if (anObj.next != null) {
                    anObj.next.prev = null;
                }
                anObj.next = null;
                anObj.parent = null;
                detection_objects.put(aBearing, null);
            }
        }
        if (this.prev != null) {
            this.prev.clearChildren();
        }
        this.prev = null;
    }

    // Correlate landmarks in cur frame to those in previous frame
    //     Where a correlation can be made with certainty, add detection object as head to a linked list
    //     containing each detection of a specific landmark over time
    public void correlateToPrev() {
        ArrayList<Double> usableBearings = new ArrayList<Double>(detection_bearings.length);
        for (double d : detection_bearings) usableBearings.add(d);

        double[] prevBearings = prev.getDetectionBearings();
        if (prevBearings == null || prevBearings.length == 0) {
            return;
        }

        double max_min_diff = 2 * Math.PI;
        double max_min_diff_bearing = 9999.0d; // value never used
        double min_diff = 2 * Math.PI;
        double min_diff_bearing = 99999.0d; //value never used
        double diff;
        while (usableBearings.size() > prevBearings.length) { // cull current landmark detections
            min_diff = 2 * Math.PI;
            for (double aBearing : usableBearings) {
                for (double aPrevBearing : prevBearings) {
                    diff = Utils.getAngleDiff(aBearing, aPrevBearing);
                    if (diff <= min_diff) {
                        min_diff = diff;
                        min_diff_bearing = aBearing;
                    }
                }
                if (min_diff >= max_min_diff) {
                    max_min_diff = min_diff;
                    max_min_diff_bearing = min_diff_bearing;
                }
            }
            usableBearings.remove(max_min_diff_bearing);
        }

        if (usableBearings.size() == 0) {
            return;
        }

        while (max_min_diff >= THRESHOLD1) { // cull landmark detections too different from any prev detections
            min_diff = 2 * Math.PI;
            for (double aBearing : usableBearings) {
                for (double aPrevBearing : prevBearings) {
                    diff = Utils.getAngleDiff(aBearing, aPrevBearing);
                    diff = Utils.getAngleDiff(aBearing, aPrevBearing);
                    if (diff <= min_diff) {
                        min_diff = diff;
                        min_diff_bearing = aBearing;
                    }
                }
                if (min_diff >= max_min_diff) {
                    max_min_diff = min_diff;
                    max_min_diff_bearing = min_diff_bearing;
                }
            }
            if (max_min_diff >= THRESHOLD1) {
                usableBearings.remove(max_min_diff_bearing);
            }
        }

        if (usableBearings.size() == 0) {
            return;
        }

        int numRows = prevBearings.length;
        int numCols = numRows;
        int[][] costMatrix = new int[numRows][numCols];

        double assignee;
        double assignment;
        for (int i = 0; i < numRows; i++) {
            for (int j = 0; j < numCols; j++) {
                if (j < usableBearings.size()) {
                    assignment = prevBearings[i];
                    assignee = usableBearings.get(j);
                    costMatrix[i][j] = (int) (Utils.getAngleDiff(assignee, assignment) / (2 * Math.PI)) * (Integer.MAX_VALUE / 2); // convert angle diff to an integer score
                } else { // to make the cost matrix square, insert placeholder value of max cost
                    costMatrix[i][j] = Integer.MAX_VALUE / 2;
                }
            }
        }
        HungarianAlgorithm ha = new HungarianAlgorithm(costMatrix);
        int[][] assignments = ha.findOptimalAssignment();
        for (int i = 0; i < usableBearings.size(); i++) { // find assignment for only non-placeholder columns
            int[] anAssignment = assignments[i];
            assignee = usableBearings.get(anAssignment[0]);
            assignment = prevBearings[anAssignment[1]];
            ObjDetection curDetection = detection_objects.get(assignee);
            ObjDetection prevDetection = this.prev.getDetectionObjects().get(assignment);
            curDetection.prev = prevDetection;
            prevDetection.next = curDetection;
        }
    }



    public void triangulate_all_objs() {
        ArrayList<ObjDetection> allObjs = get_all_objs();

        for (ObjDetection anObj : allObjs) {
            do_triangulation(anObj);
        }
    }

    public ArrayList<ObjDetection> get_all_objs() {
        ArrayList<ObjDetection> allObjs = new ArrayList<ObjDetection>(getDetectionBearings().length);
        Set set = getDetectionObjects().entrySet();
        Iterator i = set.iterator();
        while (i.hasNext()) {
            Map.Entry me = (Map.Entry)i.next();
            allObjs.add((ObjDetection) me.getValue());
        }
        return allObjs;
    }

    public void do_triangulation(ObjDetection head) {
        if (head.prev == null) {
            return;
        }
        ObjDetection cur = head.prev;
        ArrayList<ObjDetection> usablePrecursors = new ArrayList<ObjDetection>();
        int lengthAtCur = 1;
        do {
            if (lengthAtCur < MAXLISTLENGTH) {
                if (Utils.getAngleDiff(head.get_reverse_bearing(), cur.get_reverse_bearing()) >= THRESHOLD2) {
                    usablePrecursors.add(cur);
                }
            } else {
                cur.prev.next = null; // de-alloc all LL nodes beyond max allowed length
                cur.prev = null;
                cur.parent = null;
            }
            cur = cur.prev;
            lengthAtCur++;
        } while(cur != null);

        if (usablePrecursors.size() == 0) {
            return;
        }

        double c; // meters
        double B; // radians
        double C; // radians

        double b; // meters

        ArrayList<Point> triangulations = new ArrayList<Point>(usablePrecursors.size());
        double x;
        double z;

        for (ObjDetection anObj : usablePrecursors) {
            B = Utils.getAngleDiff(anObj.getParent().bearingToFrame(head.getParent()), anObj.get_detected_bearing());
            C = Utils.getAngleDiff(anObj.get_reverse_bearing(), head.get_reverse_bearing());
            c = anObj.getParent().distToFrame(head.getParent());

            b = c * Math.sin(B) / Math.sin(C);

            x = x() + b * Math.cos(head.get_detected_bearing());
            z = z() + b * Math.sin(head.get_detected_bearing());
            triangulations.add(new Point(x, z));
        }

        head.centroid = centerOf(triangulations); // write estimated location back to ObjDetection in this frame
    }

    public Point centerOf(ArrayList<Point> points) {
        double xavg = 0.0d;
        double zavg = 0.0d;
        for (Point aPoint : points) {
            xavg += aPoint.x();
            zavg += aPoint.z();
        }
        xavg /= points.size();
        zavg /= points.size();
        return new Point(xavg, zavg);
    }

}

File: src/frame.cpp
#include "frame.hpp"

#include "hungarian.hpp"
#include "stdio.h"


namespace motion_parallax {

uint64_t Frame::next_frame_number = 0;

    Frame::Frame(double camera_bearing, double x, double y, std::vector<double> relative_bearings, Frame* prev)
    : camera_global_pos_(x, y), detections_(), prev(prev), frame_number_() {
    frame_number_ = next_frame_number;
    next_frame_number++;

    if (frame_number_ >= MAX_LIST_LENGTH) {
        Frame* cur = this;
        while (cur->prev->prev != nullptr) {
            cur = cur->prev;
        }

        Frame* last = cur->prev;
        cur->prev = nullptr;
        delete last;
    }

    camera_global_pos_ = Point(x, y);
    for (int i = 0; i < relative_bearings.size(); i++) {
        double d = relative_bearings[i];

        // convert to absolute bearings
        detections_.emplace_back(this, utils::normalized(camera_bearing + d), nullptr);
    }
}

Frame::~Frame() {
    for (auto& det : detections_) {
        if (det.next != nullptr) {
            det.next->prev = nullptr;
        }
    }
}

double Frame::get_x() const {
    return camera_global_pos_.x;
}

double Frame::get_y() const {
    return camera_global_pos_.y;
}

std::vector<ObjDetection>& Frame::detections() {
    return detections_;
}

double Frame::bearingtoframe(const Frame& other) {
    double deltaz = other.get_y() - get_y();
    double deltax = other.get_x() - get_x();
    return atan2(deltaz, deltax);
}

double Frame::disttoframe(const Frame& other) {
    double deltaz = other.get_y() - get_y();
    double deltax = other.get_x() - get_x();
    return sqrt(deltaz*deltaz + deltax*deltax);
}

// correlate landmarks in cur frame to those in previous frame
//     where a correlation can be made with certainty, add detection object as head to a linked list
//     containing each detection of a specific landmark over time
void Frame::correlate_to_prev() {
    std::vector<ObjDetection>& prev_detections = prev->detections();
    if (prev_detections.empty()) {
        return;
    }

    std::vector<ObjDetection*> usable_detections;
    for (auto& det : detections_) {
        usable_detections.push_back(&det);
    }

    double max_min_diff = 0.0;
    std::vector<ObjDetection*>::iterator max_min_diff_it;
    double min_diff = 2 * M_PI;
    std::vector<ObjDetection*>::iterator min_diff_it;
    double diff;
    while (usable_detections.size() > prev_detections.size()) { // cull current landmark detections
        for (auto uit = usable_detections.begin(); uit != usable_detections.end(); uit++) {
            min_diff = 2 * M_PI;
            double usable_det_bearing = (*uit)->detected_bearing();
            for (auto pit = prev_detections.begin(); pit != prev_detections.end(); pit++) {
                double prev_det_bearing = pit->detected_bearing();
                diff = utils::get_angle_diff(usable_det_bearing, prev_det_bearing);
                if (diff <= min_diff) {
                    min_diff = diff;
                    min_diff_it = uit;
                }
            }
            if (min_diff >= max_min_diff) {
                max_min_diff = min_diff;
                max_min_diff_it = min_diff_it;
            }
        }

        usable_detections.erase(max_min_diff_it);
    }

    if (usable_detections.empty()) {
        return;
    }

    // cull landmark detections too different from any prev detections
    while (max_min_diff >= CORRELATION_ANGLE_THRESHOLD) {
        for (auto uit = usable_detections.begin(); uit != usable_detections.end(); uit++) {
            min_diff = 2 * M_PI;
            double usable_det_bearing = (*uit)->detected_bearing();
            for (auto pit = prev_detections.begin(); pit != prev_detections.end(); pit++) {
                double prev_det_bearing = pit->detected_bearing();
                diff = utils::get_angle_diff(usable_det_bearing, prev_det_bearing);
                if (diff <= min_diff) {
                    min_diff = diff;
                    min_diff_it = uit;
                }
            }
            if (min_diff >= max_min_diff) {
                max_min_diff = min_diff;
                max_min_diff_it = min_diff_it;
            }
        }

        if (max_min_diff >= CORRELATION_ANGLE_THRESHOLD) {
            usable_detections.erase(max_min_diff_it);
        }
    }

    if (usable_detections.empty()) {
        return;
    }

    int num_rows = usable_detections.size();
    int num_cols = prev_detections.size();
    std::vector<std::vector<double>> cost_matrix;

    for (ObjDetection* assignee : usable_detections) {
        auto& row = cost_matrix.emplace_back();
        for (ObjDetection& assignment : prev_detections) {
            row.push_back(utils::get_angle_diff(assignee->detected_bearing(),
                                                assignment.detected_bearing()));
        }
    }

    std::vector<int> assignments;


    //printf("[\n");
    for (auto row : cost_matrix) {
        //printf("[ ");
        for (auto el : row) {
            //printf("%f ", el);
        }
        //printf(" ]\n");
    }
    //printf("]\n");


    HungarianAlgorithm ha;
    ha.Solve(cost_matrix, assignments);

    //printf("[ ");
    for (auto n : assignments) {
        //printf("%d ", n);
    }
    //printf("]\n");

    // find assignment for only non-placeholder columns
    for (int i = 0; i < usable_detections.size(); i++) {
        ObjDetection* assignee = usable_detections[i];
        ObjDetection& assignment = prev_detections[assignments[i]];

        assignee->prev = &assignment;
        assignment.next = assignee;
    }
}



void Frame::triangulate_all_objs() {
    for (auto& det : detections_) {
        do_triangulation(&det);
    }
}

void Frame::do_triangulation(ObjDetection* head) {


    if (head == nullptr || head->prev == nullptr) {
        return;
    }

    ObjDetection* cur = head->prev;
    std::vector<ObjDetection*> usable_precursors;
    int length_at_cur = 1;
    do {
        if (length_at_cur < MAX_LIST_LENGTH) {
            if (utils::get_angle_diff(head->reverse_bearing(),
                                    cur->reverse_bearing())
                >= CONE_REVERSE_ANGLE_THRESHOLD) {
                usable_precursors.push_back(cur);
            }
        } else {
            cur->prev = nullptr;
        }
        cur = cur->prev;
        length_at_cur++;
    } while (cur != nullptr);

    //printf("got here\n");

    if (usable_precursors.empty()) {
        return;
    }

    double c; // meters
    double B; // radians
    double C; // radians

    double b; // meters

    std::vector<Point> triangulations;
    double x;
    double y;

    for (ObjDetection* det : usable_precursors) {
        B = utils::get_angle_diff(det->parent_ptr()->bearingtoframe(*(head->parent_ptr())),
                                det->detected_bearing());
        C = utils::get_angle_diff(det->reverse_bearing(),
                                head->reverse_bearing());
        c = det->parent_ptr()->disttoframe(*(head->parent_ptr()));

        b = c * sin(B) / sin(C);

        x = get_x() + b * cos(head->detected_bearing());
        y = get_y() + b * sin(head->detected_bearing());
        triangulations.emplace_back(x, y);
    }

    // write estimated location back to ObjDetection in this frame
    //printf("got here\n");
    head->centroid = center_of(triangulations);
}

Point Frame::center_of(std::vector<Point>& points) {
    double xavg = 0.0;
    double yavg = 0.0;
    for (Point& point : points) {
        xavg += point.x;
        yavg += point.y;
    }
    xavg /= points.size();
    yavg /= points.size();
    return Point(xavg, yavg);
}

}
